bundles/io/org.eclipse.smarthome.io.net/src/main/java/org/eclipse/smarthome/io/net/http/internal/SecureHttpClientFactory.java

```diff
@@ -0,0 +1,171 @@
+        logger.info("creating httpClient for endpoint {}", endpoint);
+        SslContextFactory sslContextFactory = new SslContextFactory();
+        sslContextFactory.setEndpointIdentificationAlgorithm("HTTPS");
+        if (endpoint != null && customTrustmanagerProvider != null) {
+            Stream<TrustManager> trustManagerStream = customTrustmanagerProvider.getTrustManagers(endpoint);
+            TrustManager[] trustManagers = trustManagerStream.toArray(TrustManager[]::new);
+            if (trustManagers.length > 0) {
+                logger.info("using custom trustmanagers (certificate pinning) for httpClient for endpoint {}",
+                        endpoint);
+                try {
+                    SSLContext sslContext = SSLContext.getInstance("TLS");
+                    sslContext.init(null, trustManagers, null);
+                    sslContextFactory.setSslContext(sslContext);
+                } catch (Exception ex) {
+                    throw new RuntimeException("Cannot create an TLS context for the endpoint '" + endpoint + "'!", ex);
+                }
+            }
+        }
+        String excludeCipherSuites[] = { "^.*_(MD5)$" };
+        sslContextFactory.setExcludeCipherSuites(excludeCipherSuites);
+
+        HttpClient httpClient = new HttpClient(sslContextFactory);
+        httpClient.setExecutor(threadPool);
+
+        try {
+            httpClient.start();
+        } catch (Exception e) {
+            logger.error("Could not start jetty client", e);
+            throw new RuntimeException("Could not start jetty client", e);
+        }
+
+        return httpClient;
+    }
+
+    private void checkConsumerName(String consumerName) {
+        Objects.requireNonNull(consumerName, "consumerName must not be null");
+        if (consumerName.length() < MIN_CONSUMER_NAME_LENGTH) {
+            throw new IllegalArgumentException(
+                    "consumerName " + consumerName + " too short, minimum " + MIN_CONSUMER_NAME_LENGTH);
+        }
+        if (consumerName.length() > MAX_CONSUMER_NAME_LENGTH) {
+            throw new IllegalArgumentException(
+                    "consumerName " + consumerName + " too long, maximum " + MAX_CONSUMER_NAME_LENGTH);
+        }
+        if (!CONSUMER_NAME_PATTERN.matcher(consumerName).matches()) {
+            throw new IllegalArgumentException(
+                    "consumerName " + consumerName + " contains illegal character, allowed only [a-zA-Z0-9_-]");
+        }
+    }
+
+    @Reference(service = CustomTrustManagerProvider.class, cardinality = ReferenceCardinality.OPTIONAL, policy = ReferencePolicy.DYNAMIC)
+    protected void bindCustomTrustmanagerProvider(CustomTrustManagerProvider customTrustmanagerProvider) {
+        this.customTrustmanagerProvider = customTrustmanagerProvider;
+    }
+
+    protected void unbindCustomTrustmanagerProvider(CustomTrustManagerProvider customTrustmanagerProvider) {
```

Comment:
You need to check the following:

```java
if (this.customTrustmanagerProvider == customTrustmanagerProvider) {
      this.customTrustmanagerProvider = null;
}
```

This is primarily required because of dynamic reference types, unbind methods can be called before bind methods.